### DISCUSSION TITLE
Location orientation from compass

### DISCUSSION THREAD

**Discussion by kds112** [devemux86/cruiser#625] [Q&A] [answered]
Since there is no map orientation by compass in the normal map view I have a wish for a new selection.
The direction indicator at the position marker shows the (last known) gps direction, right? Is it possible to show the (compass based) heading of my smartphone as well? Maybe a long tap on the compass to have such a selection?
As a hiker I normally switch off the smartphone for battery reasons (I don't want to carry a heavy powerbank with me). When I have to decide which direction to walk, I switch on the smartphone and I want to know where I am and which direction my phone is aligned on the north-directed map.
Thank you for considering this feature 🙂

**Comment by devemux86** [comment]
@kds112 thanks for the suggestion.

Do you mean the compass to rotate showing the north in normal map view?

**Reply by kds112** [reply]
No, I mean the (blue) direction indicator at the current map position. See screenshot 
![IMG_20240829_102121](https://github.com/user-attachments/assets/882b5304-c377-439b-b770-050281001bea)

**Reply by kds112** [reply]
This blue arrow should indicate the orientation of my smartphone... In the moment it seems to be the last known gps direction, because I am standing still for a while.

**Reply by devemux86** [reply]
@kds112 

This already works in follow location mode: long press compass to select GPS or Compass for map orientation.

We could probably extend it in map view too. And compass could also rotate to the true north.

**Reply by kds112** [reply]
It is not necessary to rotate the map. Just rotate the blue direction indicator at the map position in order to show in which direction I hold my smartphone.
As an example in the screenshot I would look in western direction with my smartphone in my hand.

**Reply by devemux86** [reply]
@kds112 

So, an option in normal map view to have the location arrow show the direction
based on GPS (current implementation) or compass.

**Reply by kds112** [reply]
Exactly 👍

**Reply by 0709wiwiwi** [reply]
Provided regular calibration, away from magnets (bookcase closure) and mounted not too near to **iron** handlebars the compass can be indicative **IF sufficiently filtered**. See [video](https://youtu.be/JwXrtD8mYD0) shows a result by a phone mounted on the handlebar. GPS direction switches to compass at speed < xx kmh. Stationary, the map direction automatically follows the handlebar direction. MTB'ers sometimes use this. Since recalibration has been necessary all too often, I still hardly use that option.

**Reply by kds112** [reply]
But if you are hiking you don't have these problems with handlebars... 🙃 And I hope that Cruiser is not a biker-only app... 🏃

**Reply by 0709wiwiwi** [reply]
I am by no means discouraging the use of the feature anyway ;-) 
Pay attention to the pain points mentioned it functions fine. 
Especially avoid these bookcases with **magnetic closure** !

**Reply by kds112** [reply]
I just use a plastic protection cover, therefore I am safe. 😇

**Comment by kds112** [comment]
Thanks a lot for the excellent updates b12 and b13, Cruiser works very well while hiking! 👍Any chance for implemetation of this small feature? I could test it during autumn hiking tours at the next weekends...

**Reply by devemux86** [reply]
@kds112 thanks for the kind words!

It is usually difficult to estimate time for new features.
But it's already in my list to check it. :slightly_smiling_face:

**Comment by devemux86** [comment]
As the code already exists in Cruiser, it is not difficult to implement and test it in the next update.

We can have a new orientation menu when long press compass in normal map view with options:
- **GPS** (default)
It works like now, the location arrow's direction is the GPS location's heading.
- **Compass**
The location arrow rotates in the direction of the compass based heading of the device.
(also using the orientation of the device: vertical, horizontal, normal / reverse)

For the compass function you must activate the "Settings | Map | Compass sensor".

---

This can be combined nicely with:
- https://github.com/devemux86/cruiser/discussions/628

**Reply by kds112** [reply]
Great! I will test whether the filtering of the compass signal is smooth enough...

One question: Can Cruiser get Information from Android about compass accuracy? Could you display high/medium/low or green/yellow/red? Then the user can recalibrate by moving the phone along 3 axis...

**Reply by devemux86** [reply]
The code already exists in Cruiser platform from other projects, I just have to enable it.
There is already a filtering to avoid jumps, we will see how smooth it is on each device.

Compass calibration? There is not something ready.
For calibration it is often better to use dedicated apps.

**Reply by kds112** [reply]
It was just the question whether information about the current compass accuracy is easily accessable from Android system and could be displayed in Cruiser... 🤔Recalibration can be done in other Apps, that's right 👍

**Reply by bios4** [reply]
Could the “compass accuracy”, for example, be displayed as a circular indicator around the compass symbol!??

Good accuracy: green circle
Medium accuracy: orange (partial) circle
Poor accuracy: red (partial) circle

Examples:
<img src="https://github.com/user-attachments/assets/d2035c94-dba7-41b5-bd3c-3b3d32457b48" width="25%">
<img src="https://github.com/user-attachments/assets/fdca6933-32d8-4147-9164-c6dd7c9b61ee" width="25%">

**Reply by bios4** [reply]
An explanation for your thumbs down would be nice @WolfgangZ42, otherwise it's difficult to classify... 😒

**Reply by devemux86** [reply]
I don't know if Android API provides compass "accuracy" info, like GPS accuracy.
Also the compass is not full screen like in compass apps to display too much data.

Let's start with the basics: use the compass sensor for direction or true north
and we will see if it works reliably and needs more improvements.

**Reply by bios4** [reply]
> I don't know if Android API provides compass "accuracy" info, like GPS accuracy.

Huh?
I made the suggestion because of this statement from you:

> The code already exists in Cruiser platform from other projects, I just have to enable it.

Anyway, it was just a suggestion.
I personally don't need that functionality at all...

**Reply by devemux86** [reply]
@bios4 

The code already in Cruiser is to use the compass sensor for directional purposes.
Compass "accuracy" (if it actually exists) is something else.

**Reply by WolfgangZ42** [reply]
> explanation for your thumbs down would be nice

(Thumbs down should be explained? Thumbs up aren't explained either. 🤔)
I just don't like additional colours - and I would feel overinformed. 
If the compass accuracy is relevant, there are lot of special apps for that - it doesn't count to the core business of a navigation app. My opinion ...

**Reply by kds112** [reply]
Well, when I am hiking with my navigation app, compass and GPS accuracy are important information and are core business of my navigation app. Please, Cruiser is not designed for motorbikers only. At least I hope so...

**Reply by WolfgangZ42** [reply]
Certainly compass and GPS are important for navigation - but not their accuracy, and that's what I mean.  
I've guided a lot of hikes with Komoot, few with Outdooractive, none of these apps shows any accuracy. 

> @kds112 ... Cruiser works very well while hiking! 👍

That's true! I did already two of my last tours via Cruiser - sucessfully.

**Reply by kds112** [reply]
Let's wait for the next beta version... Then Emux can decide to integrate more compass information or not... 🧭

**Comment by umpfmpf** [comment]
I ask myself whether the Qi charger or the electrics on the motorcycle could possibly disrupt the compass?!

**Reply by devemux86** [reply]
@umpfmpf 

You can already test this with a dedicated GPS / compass app like the [GPS Status](https://play.google.com/store/apps/details?id=com.eclipsim.gpsstatus2).

**Comment by devemux86** [comment]
Location direction from compass in normal map view is available in the latest:
- https://github.com/devemux86/cruiser/discussions/610

Activate with "Settings | Map | Compass sensor" and long press compass.
For more instructions please see [above](https://github.com/devemux86/cruiser/discussions/625#discussioncomment-10544949).

**Reply by kds112** [reply]
It seems to work great!! I will test it next weekends hiking around... 🥾

**Comment by kds112** [comment]
I come back to compass accuracy because it might be a useful Information for the user. I found the following link:
https://stackoverflow.com/questions/53418627/android-sensor-how-to-get-the-accuracy-at-anytime
Might this be useful for you? If the compass is activated in the settings, a yellow/red exclamation mark (or something else) could be displayed Inside the compass in order to indicate poor compass accuracy. The calibration itself could be done in an external app (like GPS Info), but I don't want switch all the time to GPS Info just to know that the compass accuracy is good or not...
Might this be an useful addition to the compass feature?

**Reply by devemux86** [reply]
Thanks, let's first see how it works with current implementation
and if we really need something else to not complicate the interface.

**Comment by devemux86** [comment, marked as answer]
- Location direction from compass in normal map view is available in https://github.com/devemux86/cruiser/discussions/641

Activate with "Settings | Map | Compass sensor" and long press compass.
For more instructions please see [above](https://github.com/devemux86/cruiser/discussions/625#discussioncomment-10544949).

**Comment by kds112** [comment]
This new compass function works very well 👍My compass signal is still quite jittery and needs stronger filtering. Is it possible to get this option as a selection or a permanent implemetation if other smartphones have the same effect? 🧭

**Reply by devemux86** [reply]
@kds112 Thanks for the feedback.

Could be related to the device, as here the compass seems to work quite smoothly.
We can think of more improvements. Have you calibrated it?

How other users see the compass on their devices?

**Reply by kds112** [reply]
It is fully calibrated and still dithering a few degrees several times per second. Maybe it would be enough to update the compass two times per second or so...

**Reply by kds112** [reply]
I have this effect of a jittery compass indicator on two devices, Realme 8 (Android 13) and Samsung A55 (Android 14). Anybody else?

**Reply by Ciutnik** [reply]
This is indeed happening on my motorcycle Cruiser. Tested on a flat surface with no metal nearby.
I didn't notice it before because I didn't turn on the compass. The question is whether it is very bothersome and whether software filtering of this (i.e. additional work and phone resources) will not be more troublesome than this effect? 🙂

https://github.com/user-attachments/assets/03bbd2ca-f7b7-4278-97f3-b873a858623e

**Reply by kds112** [reply]
It is exactly the effect I have on my smartphone. 👍 You are right I can live with it. Just in case stronger filtering can be done without much work, Emux could implement "something".

**Reply by devemux86** [reply]
@Ciutnik Thanks for the video with both functions. This is needed to understand what you see.

I will check how more filtering is possible in next update and see how it appears on your devices.

**Reply by 0709wiwiwi** [reply]
The reference is sure Locus map compass.
See compare video with the previous Kurviger version.
Notice the stable Locus compass direction indication. 
https://github.com/devemux86/cruiser/discussions/625#discussioncomment-10486484

**Reply by devemux86** [reply]
**Cruiser 5.0.1-b1** is available with improved compass filtering. Can you check it?
- https://github.com/devemux86/cruiser/discussions/641#discussioncomment-10956572

**Reply by Ciutnik** [reply]
Thanks for the quick release! But sorry, the compass seems to be shaking just like before...

**Reply by devemux86** [reply]
Have you calibrated the compass?
Have you tried the compass with other apps?

Here even without calibration, both old and new versions are very smooth.
So this is clearly related to the device's compass sensor quality.

I could make the compass to update less frequently, like 1-2 times per second,
but this will not improve the device compass sensor hardware. :slightly_smiling_face:

**Reply by kds112** [reply]
Well, for me it is a big improvement 👍The compass is now very smooth.

**Reply by Ciutnik** [reply]
It's hard to say what's causing it. The video I posted is from Samsung. But it looks exactly the same on Pixel 1. Or is it external interference? I'm in the middle of the city...
At the same time, I confirm the observation of @0709wiwiwi - I don't see such an effect in Locus.

**Reply by Ciutnik** [reply]
OK, I watched my video again, and I have to admit it's better now! Although the compass still shakes, it's not as noticeable. I apologize to @devemux86 for misleading you.

**Reply by devemux86** [reply]
@Ciutnik Is it possible to post a similar video on same device with the new version to compare them?

**Reply by Ciutnik** [reply]
_oh, I posted the answer in the wrong place:_ https://github.com/devemux86/cruiser/discussions/625#discussioncomment-10957955

**Comment by Ciutnik** [comment]
Of course. I'm posting a video from 10 minutes ago. Same device, same place, 5.0.1-b1:

https://github.com/user-attachments/assets/9d63ed8d-634c-43da-b912-fabe92964052

**Reply by devemux86** [reply]
Please try with **5.0.1-b2** which has more filtering and refresh every 1 sec.

**Reply by Ciutnik** [reply]
Yes, now (b2) the compass is calm 🙂👍
I don't know if refreshing every second isn't too infrequent for a walker? Maybe half a second would be smoother? I don't think it matters in any vehicle.

**Comment by kds112** [comment]
B1 was definitly better. The refresh rate 1 sec is too slow. At least twice per second is necessary. B2 compass is too jerky right now.

**Reply by Ciutnik** [reply]
I agree with the above opinion @kds112

**Comment by devemux86** [comment]
Please try with **5.0.1-b3**, it has the same increased filtering as b2, but the refresh is 2 times per sec.

**Reply by kds112** [reply]
I think, B1 was still better than B3. Is it possible to have the stronger filtering without restriction of the refresh rate?

**Reply by devemux86** [reply]
> Is it possible to have the stronger filtering without restriction of the refresh rate?

It is possible, we have to make tests until we find a nice middle solution.

**Comment by devemux86** [comment]
Please try with **5.0.1-b4**, the refresh is every 200 ms.

**Reply by kds112** [reply]
It looks like a good compromise👍

**Reply by Ciutnik** [reply]
It's more than good. The compass is like set in concrete 🙂
I wonder if 200 ms filtering is even necessary?

**Reply by devemux86** [reply]
We can try more frequent refresh for smoother movement.

**Reply by Ciutnik** [reply]
@devemux86 Does the refresh rate affect the battery or overall performance? Because if so, then in my opinion there is no need to increase the rate.

**Reply by devemux86** [reply]
> Does the refresh rate affect the battery or overall performance?

Not really.

Note that the GPS location with all its properties, such as heading, speed, etc., is updated at best every second.

That's why navigators cannot refresh faster than one time per second.
Only animations can help with worse battery and then the animated location on screen is not the real location.

**Comment by devemux86** [comment]
Please try **5.0.1-b5**.

**Reply by kds112** [reply]
Perfect compass movement 🎉🏁

**Reply by Ciutnik** [reply]
I confirm! It's perfect. The compass doesn't shake and turns very smoothly. Thanks! 👍
Instead of a video, I could upload a photo - it would be the same 🙂

**Reply by Ciutnik** [reply]
I don't think it's bad, but I hadn't really thought about it before. While you cannot actually see the movement of the compass arrow, you can see the movement in the map view (it's just an obvious matter of size).

But I really don't think it's a problem, I just want to inform you about this behavior:

https://github.com/user-attachments/assets/906e54db-723f-40c9-bc5f-4c1fbe94749c

**Comment by kds112** [comment]
Just for curiosity reasons as asked earlier in this discussion: the compass accuracy... 😇 Might it possible to catch the Android Information? And in case of bad accuracy displaying a yellow or red spot Inside the compass icon?

**Reply by devemux86** [reply]
Compass will not be easily readable if we add many colors and info in it, except the mandatory north direction.
We will see, need to think about it.

**Reply by kds112** [reply]
Maybe a panel like the gps Info? 🙂
![IMG_20241017_103909](https://github.com/user-attachments/assets/168a0381-2782-4e9c-b063-dffefd7fc858)

Then it could be disabled by someone who does not use the compass functionality.

**Reply by devemux86** [reply]
A panel option for such info is indeed a better solution.

**Comment by devemux86** [comment]
> you can see the movement in the map view

Please try **5.0.1-b6**.

**Reply by Ciutnik** [reply]
B6. It seems to me that nothing has changed:

https://github.com/user-attachments/assets/00a6819d-4e84-4008-8442-3a8316d9d0c6

**Reply by kds112** [reply]
B6 seems to be more or less like b5... Are you using different Filters?

**Reply by devemux86** [reply]
Compass icon should be ok since **b5**.
We now look at the map when follow mode is set to compass, as reported in the above videos.

**Reply by Ciutnik** [reply]
I thought the compass and map use exactly the same processing and filtering engine. And the visibility of movement on the map is only due to it size compared to a compass needle or a tiny direction arrow... 🙂

**Reply by devemux86** [reply]
It should be, so maybe we need to improve something there.

**Reply by kds112** [reply]
For me it is really ok, strong filtering, still lively compass and the map movement is acceptable.

**Reply by devemux86** [reply]
We mean follow mode enabled and then select "Compass direction" as "Map orientation" (long press compass).

**Reply by kds112** [reply]
Yes I know... Latest beta:

https://github.com/user-attachments/assets/4964d8ca-b78c-4cb7-8303-8cb7d0ac8a84

**Reply by devemux86** [reply]
@Ciutnik @kds112 

At first I thought the report was that the compass was fixed but the map was moving.
Checking the videos again carefully, the compass is also moving, so all should be fine.

Or is there anything else we need to improve?
The map follows the compass movement, so is the compass now smooth and ok?

**Reply by kds112** [reply]
Yes, everything looks convincing👍

**Reply by kds112** [reply]
My last wish is the accuracy... 😁 Next week I will hike in South Tyrol and Cruiser can show its power 💪

**Comment by devemux86** [comment]
**Cruiser 5.0.1-b7** is available with compass internal improvements (no functionality change).

**Reply by Ciutnik** [reply]
Well, I guess there's no difference from the b6. While the compass seems to be stationary (I know, it just looks that way), the map rotates a bit in follow mode.

**Reply by devemux86** [reply]
If we maximize the video and look the compass, it also moves with the map.

Is the device stable on a surface?
Because then the compass / map should not be moved?
Isn't this supposed to be fixed with these betas?

**Reply by Ciutnik** [reply]
> Is the device stable on a surface?

Yes, of course, it lies flat on a wooden table, away from anything metal or electrical.

As a curiosity: I once walked across a large bridge in the city with a tourist GPS (Garmin Etrex) in my hand. About halfway along the bridge, a several dozen meters above the river, there was a place - a section of a few meters - where the three-axis compass went completely crazy! The readings turned 180°! I think something beneath the surface was causing a very strong electric field.
Maybe my close neighbors have something like that? 😃

**Reply by kds112** [reply]
Map rotation is smooth enough. The current map moves are okay, the cheap compass units inside the smartphones won't deliver better results... 🧭

**Comment by devemux86** [comment]
More filtering improvements are included in **Cruiser 5.0.1-b8**:
- https://github.com/devemux86/cruiser/discussions/641#discussioncomment-10982507

**Reply by kds112** [reply]
Still looking good! 👍

**Comment by devemux86** [comment]
- Compass sensor filtering was improved in https://github.com/devemux86/cruiser/discussions/641
